instrumentation/grpc/lib/opentelemetry/instrumentation/grpc/patches/active_call.rb

```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module GRPC
+      module Patches
+        module ActiveCall
+          def initialize(call, marshal, unmarshal, deadline, started: true, metadata_received: false, metadata_to_send: nil)
+            super
+            @span = OpenTelemetry::Trace.current_span
+          end
+
+          # We need a setter because while we can easily capture a correct trace during initialization
+          # on the client side, it's a little harder in the accept loop on the server. So, we provide
+          # a way for server-side ActiveCalls to set the span after the ActiveCall is created.
+          def span=(span)
+            @span = span
+          end
+
+          # Handles the "unary" client call.
+          def request_response(req, metadata: {})
+            OpenTelemetry::Trace.with_span(@span) { super }
+          end
+
+          # Handles the "client streaming" client call
+          def client_streamer(requests, metadata: {})
+            OpenTelemetry::Trace.with_span(@span) { super }
+          end
+
+          # Handles the "server streaming" client call
+          def server_streamer(req, metadata: {})
+            OpenTelemetry::Trace.with_span(@span) { super }
+          end
+
+          # Handles the "bi-directional streaming" client call
+          def bidi_streamer(requests, metadata: {})
+            OpenTelemetry::Trace.with_span(@span) { super }
+          end
+
+          # There are some instances where the client may receive
+          # an enumerator back from a streaming call, and in those cases
+          # the actual calls to read from the server are not executed until
+          # the client starts calling `next`. So we must ensure our desired
+          # span is active here, for those cases.
+          def each_remote_read_then_finish
+            OpenTelemetry::Trace.with_span(@span) { super }
+          end
+
+          # This method is locked with a mutex and only called
+          # from set_{input,output}_stream_done. We use it to finish
+          # off the span we've been tracking throughout the
+          # lifecycle of this ActiveCall.
+          def maybe_finish_and_close_call_locked
+            super
+
+            if @call_finished && @span.recording?
+              unless status.nil?
+                @span.set_attribute("rpc.grpc.status_code", status.code)
+                @span.status = OpenTelemetry::Trace::Status.error unless status.code == ::GRPC::Core::StatusCodes::OK
```

Comment:
IIRC we recently discussed that auto-instrumentation should not set the `span.status`. Is that correct?

Reply:
IIRC the outcome was that yes, generally speaking instrumentation shouldn't set the status (and certainly should not set to "OK", which is now a "final" state that can't be changed). But, I think errors _can_ still be set.

I was working from the [relevant part of the RPC semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/rpc.md#grpc-status), which state:

> The Span Status MUST be left unset for an OK gRPC status code, and set to Error for all others.

The current [relevant portion of the API docs](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#set-status) seem to say that this is still allowable:

> The status code SHOULD remain unset, except for the following circumstances:
> When the status is set to Error by Instrumentation Libraries, the status codes SHOULD be documented and predictable. The status code should only be set to Error according to the rules defined within the semantic conventions. For operations not covered by the semantic conventions, Instrumentation Libraries SHOULD publish their own conventions, including status codes.
> Generally, Instrumentation Libraries SHOULD NOT set the status code to Ok, unless explicitly configured to do so. Instrumention libraries SHOULD leave the status code as Unset unless there is an error, as described above.

So - I think this is okay still, based on the specs. But I'm not 100% certain actually.

Reply:
The spec is verbose, but the essence is (for instrumentation):
1. Don't ever set `span.status = OpenTelemetry::Trace::Status.ok`
2. If there's an error, and the semantic conventions say you should set an error, set `span.status = OpenTelemetry::Trace::Status.error`
3. Otherwise leave `span.status` alone! What did it ever do to you anyway?!